pkg/util/dynclient/dynclient.go

```diff
@@ -0,0 +1,108 @@
+package dynclient
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	dynamic "k8s.io/client-go/deprecated-dynamic"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+type DynClient interface {
+	GetOrList(ctx context.Context, kind, namespace, name string) (*unstructured.UnstructuredList, error)
+}
+
+type dynclient struct {
+	log *logrus.Entry
+
+	cli *discovery.DiscoveryClient
+	dyn dynamic.ClientPool
+	grs []*restmapper.APIGroupResources
+}
+
+func New(log *logrus.Entry, restconfig *rest.Config) (DynClient, error) {
+	d := &dynclient{
+		log: log,
+	}
+
+	var err error
+	d.cli, err = discovery.NewDiscoveryClientForConfig(restconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	grs, err := restmapper.GetAPIGroupResources(d.cli)
+	if err != nil {
+		return nil, err
+	}
+	d.grs = grs
+
+	rm := restmapper.NewDiscoveryRESTMapper(d.grs)
+	d.dyn = dynamic.NewClientPool(restconfig, rm, dynamic.LegacyAPIPathResolverFunc)
+
+	return d, nil
+}
+
+// contains returns true if haystack contains needle
+func contains(haystack []string, needle string) bool {
+	for _, s := range haystack {
+		if s == needle {
+			return true
+		}
+	}
+	return false
+}
+
+func (d *dynclient) GetOrList(ctx context.Context, kind, namespace, name string) (*unstructured.UnstructuredList, error) {
+	if strings.EqualFold("Secret", kind) {
+		return nil, nil
+	}
+
+	for _, gr := range d.grs {
+		for version, resources := range gr.VersionedResources {
+			for _, resource := range resources {
+				if strings.ContainsRune(resource.Name, '/') { // no subresources
+					continue
+				}
+				if !contains(resource.Verbs, "list") { // must be able to list
+					continue
+				}
+
+				gvk := schema.GroupVersionKind{Group: gr.Group.Name, Version: version, Kind: resource.Kind}
+				gk := gvk.GroupKind()
+				if !strings.EqualFold(gk.String(), kind) {
+					continue
+				}
+				dc, err := d.dyn.ClientForGroupVersionKind(gvk)
+				if err != nil {
+					return nil, err
+				}
+
+				lo := metav1.ListOptions{}
+				if name != "" {
+					lo.FieldSelector = "metadata.name=" + name
+				}
+				ro, err := dc.Resource(&resource, namespace).List(lo)
```

Comment:
does this return an empty list instead of an error when "get"ting an object?  We should have two util/dynclient functions, get and list; one front-end /admin action is OK: if name is specified, call get, if not, call list?

Reply:
i was trying to make the response consistent (metav1.List) instead of having to deal with 2 different responses

Reply:
I'd like to match the oc/kubectl behaviour - i.e. the return type of `oc get pods` is a list of pods, and the return of `oc get pod foo` is a pod, not a list of pods of length 1.